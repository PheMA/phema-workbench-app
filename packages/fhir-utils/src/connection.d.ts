export interface AuthConfig {
  basic?: string;
  oauth?: string;
}

export interface Connection {
  id: string;
  name?: string;
}

export interface Parameter {
  id: string;
  name: string;
  value: string;
}

export interface FHIRServerParameter {
  id: string;
  name: string;
  valueString: string;
}

export interface FHIRConnection extends Connection {
  fhirBaseUrl: string;
  auth?: AuthConfig;
  parameters?: FHIRServerParameter[];
}
