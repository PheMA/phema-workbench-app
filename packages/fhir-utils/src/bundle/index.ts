import { R4 } from "@ahryman40k/ts-fhir-types";

const extractResources = (
  bundle: R4.IBundle,
  resourceTypes: string[]
): R4.IDomainResource[] => {
  const resources: R4.IDomainResource[] = [];

  bundle.entry?.forEach((entry) => {
    if (resourceTypes?.includes(entry?.resource?.resourceType)) {
      resources.push(entry.resource);
    }
  });

  return resources;
};

export { extractResources };
