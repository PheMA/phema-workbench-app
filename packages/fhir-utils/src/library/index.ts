import { R4 } from "@ahryman40k/ts-fhir-types";
import { v4 as uuidv4 } from 'uuid';

const extractCqlFromLibrary = (
    library: R4.ILibrary
): string => {
    const content = library?.content.find(c => {
        return c.contentType === "text/cql";
    });

    return atob(content.data);
}

const newEmptyLibrary = () => {
    return {
        "resourceType": "Library",
        "id": uuidv4(),
        "version": "1.0.0",
        "content": [{
            "contentType": "text/cql",
            data: btoa(`/**
 * PhEMA Phenotype
 *
 * Name : Unnamed
 * ID   : 
 * Url  : 
 */
library "Unnamed" version '1.0.0'

using FHIR version '4.0.0'

include FHIRHelpers version '4.0.0' called FHIRHelpers

context Patient

define "Test":
    true`)
        }]
    }
}

export { extractCqlFromLibrary, newEmptyLibrary }