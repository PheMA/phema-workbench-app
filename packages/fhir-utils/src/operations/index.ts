export const buildParametersResource = (params) => {
    let resource = {
        resourceType: "Parameters",
        parameter: [],
    };

    params.forEach((param) => {
        resource.parameter.push({
            name: param.name,
            valueString: param.value,
        });
    });

    return resource;
};