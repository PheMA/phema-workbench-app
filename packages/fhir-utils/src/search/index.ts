import { R4 } from "@ahryman40k/ts-fhir-types";
import { FHIRConnection } from "../connection";

interface SearchParameters {
  fhirConnection: FHIRConnection;
  resourceType: string;
  parameters: {
    [parameter: string]: string;
  };
}

async function search({
  fhirConnection,
  resourceType,
  parameters,
}: SearchParameters): Promise<R4.IBundle> {
  const queryString = new URLSearchParams();

  const headers = {
    Accept: "application/json",
  };

  if (fhirConnection.auth) {
    if (fhirConnection.auth.basic) {
      headers["Authorization"] = `Basic ${fhirConnection.auth.basic}`;
    } else if (fhirConnection.auth.oauth) {
      headers["Authorization"] = `Bearer ${fhirConnection.auth.oauth}`;
    }
  }

  Object.keys(parameters).forEach((key) => {
    if (parameters[key]) {
      queryString.set(key, parameters[key]);
    }
  });

  const url = `${
    fhirConnection.fhirBaseUrl
  }/${resourceType}?${queryString.toString()}`;

  return fetch(url, { headers }).then((res) => res.json());
}

export { search };
