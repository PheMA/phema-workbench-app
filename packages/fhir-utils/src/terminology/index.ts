import { R4 } from "@ahryman40k/ts-fhir-types";

import { FHIRUtils, FHIRConnection } from "../";

interface ExpandParameters {
  fhirConnection: FHIRConnection;
  valueSetId: string;
}

async function expand({
  fhirConnection,
  valueSetId,
}: ExpandParameters): Promise<R4.IValueSet> {
  const headers = {
    Accept: "application/json",
  };

  if (fhirConnection.auth) {
    if (fhirConnection.auth.basic) {
      headers["Authorization"] = `Basic ${fhirConnection.auth.basic}`;
    } else if (fhirConnection.auth.oauth) {
      headers["Authorization"] = `Bearer ${fhirConnection.auth.oauth}`;
    }
  }

  const url = `${fhirConnection.fhirBaseUrl}/ValueSet/${valueSetId}/$expand`;

  return fetch(url, { headers }).then((res) => res.json());
}

interface ExtractValueSetDependenciesParameters {
  fhirConnection: FHIRConnection;
  valueSet: R4.IValueSet;
}

async function extractValueSetDependencies({
  fhirConnection,
  valueSet,
}: ExtractValueSetDependenciesParameters): Promise<
  (R4.IValueSet | R4.ICodeSystem)[]
> {
  const dependencies = new Set<string>();

  // collect dependencies
  valueSet?.compose?.include?.forEach((entry) => {
    if (entry.valueSet) {
      entry.valueSet.forEach((valueSetUrl) => {
        let url = valueSetUrl;

        // magic to support proxying, and because VSAC doesn't
        // support the URL search parameter ðŸ™„
        if (!valueSetUrl.startsWith(fhirConnection.fhirBaseUrl)) {
          const components = valueSetUrl.split("/");

          url = `${fhirConnection.fhirBaseUrl}/ValueSet/${components.pop()}`;
        }

        dependencies.add(url);
      });
    } else if (entry.system) {
      const queryString = new URLSearchParams();

      queryString.set("system", entry.system);

      if (entry.version) {
        queryString.set("version", entry.version);
      }

      // sort by date so we pick the latest
      queryString.set("_sort", "-date");

      dependencies.add(
        `${fhirConnection.fhirBaseUrl}/CodeSystem?${queryString.toString()}`
      );
    }
  });

  // fetch dependencies
  const dependencyResponses = await Promise.allSettled(
    [...dependencies].map((url) => FHIRUtils.fetch({ fhirConnection, url }))
  );

  const dependencyResources = [];

  for (let i = 0; i < dependencyResponses.length; i++) {
    const response = dependencyResponses[i];

    if (response.status === "fulfilled") {
      const resource = response.value;

      if ((resource as R4.IValueSet).resourceType === "ValueSet") {
        // We retrieved a ValueSet resource directly
        dependencyResources.push(resource);

        // Recursively add value set dependencies
        const recursiveDeps = await extractValueSetDependencies({
          fhirConnection,
          valueSet: resource as R4.IValueSet,
        });

        dependencyResources.push(...recursiveDeps);
      } else if ((resource as R4.IBundle).resourceType === "Bundle") {
        const bundle = resource as R4.IBundle;

        if (bundle.entry) {
          // We search for a code system by name (and maybe version)
          dependencyResources.push(bundle.entry[0]?.resource);
        }
      }
    } else {
      console.warn("Error retrieving ValueSet dependency");
    }
  }

  return dependencyResources;
}

// the following functions could be in bundle utils?

interface BundleContainsValueSetParameters {
  bundle: R4.IBundle;
  valueSet: R4.IValueSet;
}

function bundleContainsValueSet({
  bundle,
  valueSet,
}: BundleContainsValueSetParameters): boolean {
  return bundle?.entry?.some((entry) => {
    if (!entry.resource || entry.resource?.resourceType !== "ValueSet") {
      return false;
    }

    const resource = entry.resource;

    // first try url
    if (valueSet?.url && resource.url) {
      return resource.url === valueSet.url;
    }

    // next try identifier
    if (valueSet?.identifier && valueSet?.identifier[0]) {
      if (resource.identifier && resource.identifier[0]) {
        // should check system here too
        if (valueSet?.identifier[0].value === resource.identifier[0].value) {
          return true;
        }
      }
    }

    // finally try id
    // NB: This isn't actually spec compliant
    if (valueSet?.id) {
      if (valueSet?.id == resource.id) {
        return true;
      } else {
        if (resource.identifier && resource.identifier[0]) {
          if (resource.identifier[0].value.includes(valueSet?.id)) {
            return true;
          }
        }
      }
    }

    return false;
  });
}

interface BundleContainsCodeSystemParameters {
  bundle: R4.IBundle;
  codeSystem: R4.ICodeSystem;
}

function bundleContainsCodeSystem({
  bundle,
  codeSystem,
}: BundleContainsCodeSystemParameters): boolean {
  return bundle?.entry?.some((entry) => {
    const resource = entry.resource;

    if (
      resource.resourceType === "CodeSystem" &&
      resource.url === codeSystem.url
    ) {
      return true;
    }
  });
}

interface ValueSetItem {
  code: string;
  display?: string;
  system?: string;
  version?: string;
}

function valueSetToCodeArray(
  valueSet: R4.IValueSet
): ValueSetItem[] {
  const codes = [];

  valueSet?.compose?.include.forEach(include => {
    include?.concept?.forEach(concept => {
      codes.push({
        code: concept.code,
        display: concept.display,
        system: include.system,
        version: include.version
      })
    });
  });

  return codes;
}

export {
  expand,
  extractValueSetDependencies,
  bundleContainsValueSet,
  bundleContainsCodeSystem,
  valueSetToCodeArray
};
