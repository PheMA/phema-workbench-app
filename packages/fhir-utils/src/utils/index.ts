import { R4 } from "@ahryman40k/ts-fhir-types";

interface GetParameters {
  fhirConnection: FHIRConnection;
  resourceType: string;
  resourceId: string;
}

// get a resource by id
async function get({
  fhirConnection,
  resourceType,
  resourceId,
}: GetParameters): Promise<R4.IResource> {
  const headers = {
    Accept: "application/json",
  };

  if (fhirConnection.auth) {
    if (fhirConnection.auth.basic) {
      headers["Authorization"] = `Basic ${fhirConnection.auth.basic}`;
    } else if (fhirConnection.auth.oauth) {
      headers["Authorization"] = `Bearer ${fhirConnection.auth.oauth}`;
    }
  }

  const url = `${fhirConnection.fhirBaseUrl}/${resourceType}/${resourceId}`;

  return global.fetch(url, { headers }).then((res) => res.json());
}

interface FetchParameters {
  fhirConnection: FHIRConnection;
  url: string;
  method?: string;
  body?: object;
}

// simple wrapper around fetch to set headers
async function fetch({
  fhirConnection,
  url,
  method,
  body,
}: FetchParameters): Promise<R4.IResource> {
  const headers = {
    Accept: "application/json",
    "Content-Type": "application/json",
  };

  if (fhirConnection.auth) {
    if (fhirConnection.auth.basic) {
      headers["Authorization"] = `Basic ${fhirConnection.auth.basic}`;
    } else if (fhirConnection.auth.oauth) {
      headers["Authorization"] = `Bearer ${fhirConnection.auth.oauth}`;
    }
  }

  if (body && method === "GET") {
    throw new Error("Cannot perform GET with body");
  }

  const config = {
    headers,
    method: method || "GET",
    body: JSON.stringify(body),
  };

  return global.fetch(url, config).then((res) => res.json());
}

interface SubmitBundleParameters {
  fhirConnection: FHIRConnection;
  bundle: R4.IBundle;
}

async function submitBundle({
  fhirConnection,
  bundle,
}: SubmitBundleParameters): Promise<R4.IBundle> {
  return fetch({
    fhirConnection,
    url: fhirConnection.fhirBaseUrl,
    method: "POST",
    body: bundle,
  }) as Promise<R4.IBundle>;
}

export { get, fetch, submitBundle };
