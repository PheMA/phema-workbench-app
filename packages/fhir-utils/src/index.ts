import * as BundleUtils from "./bundle";

export { BundleUtils };
