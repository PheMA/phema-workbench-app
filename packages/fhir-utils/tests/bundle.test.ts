import { BundleUtils } from "../src";

const bundle = require("./fixtures/bundle.test.json");

test("Gets CodeSystems out of Bundle", () => {
  const codeSystems = BundleUtils.extractResources(bundle, ["CodeSystem"]);

  expect(codeSystems).toMatchSnapshot();
});
