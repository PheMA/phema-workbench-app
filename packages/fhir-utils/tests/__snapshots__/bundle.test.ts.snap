// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gets CodeSystems out of Bundle 1`] = `
Array [
  Object {
    "content": "complete",
    "date": "2017-09-27T00:00:00-04:00",
    "id": "ActCode",
    "identifier": Object {
      "system": "urn:ietf:rfc:3986",
      "value": "urn:oid:2.16.840.1.113883.5.4",
    },
    "meta": Object {
      "lastUpdated": "2017-11-20T00:00:00.000-05:00",
      "versionId": "3244932127",
    },
    "name": "ActCode",
    "resourceType": "CodeSystem",
    "text": Object {
      "div": "<div xmlns=\\"http://www.w3.org/1999/xhtml\\"><h2>ActCode</h2><p>This code system http://hl7.org/fhir/v3/ActCode defines the following codes:</p><table class=\\"codes\\"><tr><td><b>Code</b></td><td><b>Definition</b></td></tr></table></div>",
      "status": "generated",
    },
    "title": "ActCode",
    "url": "http://hl7.org/fhir/v3/ActCode",
    "version": "HL7V3.0_2017-07",
  },
  Object {
    "content": "complete",
    "date": "2019-12-27T00:00:00-05:00",
    "id": "ActCode",
    "identifier": Object {
      "system": "urn:ietf:rfc:3986",
      "value": "urn:oid:2.16.840.1.113883.5.4",
    },
    "meta": Object {
      "lastUpdated": "2020-01-16T00:00:00.000-05:00",
      "versionId": "1333212229",
    },
    "name": "ActCode",
    "resourceType": "CodeSystem",
    "text": Object {
      "div": "<div xmlns=\\"http://www.w3.org/1999/xhtml\\"><h2>ActCode</h2><p>This code system http://hl7.org/fhir/v3/ActCode defines the following codes:</p><table class=\\"codes\\"><tr><td><b>Code</b></td><td><b>Definition</b></td></tr></table></div>",
      "status": "generated",
    },
    "title": "ActCode",
    "url": "http://hl7.org/fhir/v3/ActCode",
    "version": "HL7V3.0_2019-12",
  },
]
`;
